### ISSUE TITLE
locale not set by domain

### ISSUE THREAD

**Issue by brasileric** [tractorcow/silverstripe-fluent#439]
Hi,
I use multiple domains which each their own language. But I still have to add the prefix after the domain for the right locale. If I do not it uses the default locale or if the default is not set in the cms it uses en_US because that's the first locale in the row. For example:

www.mydomain.be -> sets the default locale (or en_US)
www.mydomain.be/nl_BE -> sets the nl_BE locale

www.mydomain.nl -> sets the default locale (or en_US)
www.mydomain.nl/nl_NL -> sets the nl_NL locale

www.mydomain.com -> sets the default locale (or en_US)
www.mydomain.nl/en_US -> sets the en_US locale

I suppose that if I use multiple domains and just one locale per domain (set as default for that domain as well) I don't have to add the prefix, or am I wrong?

SilverStripe: 4.1.2
Fluent: tried dev-master & tag 4.1.3

have added ```SS_FLUENT_FORCE_DOMAIN=true``` to ```.env```
have set ```disable_default_prefix``` to ```true```
tried with the domains include ```www``` and exclude ```www```

Looks like issue #419, maybe this is not fixed yet?

**Comment by robbieaverill** [comment]
Fluent should automatically detect the default locale for a domain and use it, provided no other means of determining the current locale are available first. This sounds like a bug to me, thanks for reporting.

**Comment by robbieaverill** [comment]
Fixed with #452

**Comment by obj63mc** [comment]
I am currently having this same issue on a site of mine.  It has the same type of configuration but it doesn't seem to be respecting the domain name.

* domain.com - en_US as expected
* domain.com/en_GB - UK as expected
* domain.co.uk - US - should return UK
* domain.co.uk/en_GB - UK returned as expected.

YML config is below and I have added SS_FLUENT_FORCE_DOMAIN=true to .env

    ---
    Name: myfluentconfig
    After:
      - '#versionedfiles'
      - '#elemental'
      - '#silverstripe-seo'
      - '#swiftapp'
      - '#blogconfig'
    ---
    TractorCow\Fluent\Extension\FluentDirectorExtension:
      disable_default_prefix: true
    TractorCow\Fluent\Extension\FluentExtension:
      frontend_publish_required: false
    TractorCow\Fluent\Middleware\DetectLocaleMiddleware:
      persist_cookie: true
    DNADesign\Elemental\Models\ElementalArea:
      extensions:
        - 'TractorCow\Fluent\Extension\FluentVersionedExtension'
    DNADesign\Elemental\Models\BaseElement:
      extensions:
        - 'Swift\Extensions\BaseElementExtension'
    Heyday\MenuManager\MenuSet:
      extensions:
        - 'TractorCow\Fluent\Extension\FluentExtension'
    Heyday\MenuManager\MenuItem:
      extensions:
        - 'TractorCow\Fluent\Extension\FluentExtension'
    SilverStripe\SiteConfig\SiteConfig:
      extensions:
        - 'TractorCow\Fluent\Extension\FluentExtension'
    gorriecoe\Link\Models\Link:
      extensions:
        - 'TractorCow\Fluent\Extension\FluentExtension'

Screen shots of CMS config

<img width="1476" alt="Screen Shot 2020-01-22 at 7 21 14 PM" src="https://user-images.githubusercontent.com/443120/72949072-d1822e80-3d4c-11ea-8617-bfa28b807983.png">
<img width="1465" alt="Screen Shot 2020-01-22 at 7 20 57 PM" src="https://user-images.githubusercontent.com/443120/72949073-d1822e80-3d4c-11ea-9282-c4c4dcfb5434.png">
<img width="1228" alt="Screen Shot 2020-01-22 at 7 20 24 PM" src="https://user-images.githubusercontent.com/443120/72949074-d1822e80-3d4c-11ea-8400-43d104ea80c0.png">
<img width="1110" alt="Screen Shot 2020-01-22 at 7 20 10 PM" src="https://user-images.githubusercontent.com/443120/72949075-d1822e80-3d4c-11ea-9259-411fd17f85ae.png">

Some notes that I also see - in the LocaleMenu - it is only using a relative URL for the default language (US) in my case, but uses the absolute URL for the UK links and has the correct link with out the urlsegment provided.  eg.

* US link for homepage is /  should be domain.com/
* UK link for homepage is listed as domain.co.uk/ as expected

If I go to the UK site and try to go back to US language it won't work as the URL is relative instead of absolute.

Additionally, in the metatags - the alternate language references are correct though and using the correct absolute links as expected.

I am currently running fluent v4.4 and SS 4.5.0

Any ideas on how to fix this?

**Comment by tractorcow** [comment]
Could it be that you're accessing the domain without the `www.` prefix? Try setting up a www redirect (via htaccess).

**Comment by obj63mc** [comment]
No I was accessing it with www, it seems to be an issue with following configs - 

in .env

    SS_BASE_URL='https://www.domain.com/'

in my yml config - 

    SilverStripe\Control\Director:
      alternate_base_url: '`SS_BASE_URL`'

If I have those set to the US domain then it will always return the default language.  If I remove those then it works fine but then the CMS admin no longer works for me, basically all the JS/CSS won't include properly as it is being dynamically injected (not sure why but...)

Still debugging trying to find a way around this, even tried manually configuring everything in YML but that still wasn't working for me, example - 

	---
	Name: myfluentconfig
	After:
	- '#fluentconfig'
	- '#versionedfiles'
	- '#elemental'
	- '#silverstripe-seo'
	- '#swiftapp'
	- '#blogconfig'
	---
	TractorCow\Fluent\Extension\FluentDirectorExtension:
	disable_default_prefix: true
	detect_locale: false
	default_locale: en_US
	locales:
		- en_US
		- en_GB
	aliases:
		en_GB: uk
		en_US: us
	domains:
		domain.com:
		locales:
			- en_US
		domain.co.uk:
		locales:
			- en_GB
	TractorCow\Fluent\Extension\FluentExtension:
	frontend_publish_required: false
	TractorCow\Fluent\Middleware\DetectLocaleMiddleware:
	persist_cookie: true
	DNADesign\Elemental\Models\ElementalArea:
	extensions:
		- 'TractorCow\Fluent\Extension\FluentVersionedExtension'
	DNADesign\Elemental\Models\BaseElement:
	extensions:
		- 'Swift\Extensions\BaseElementExtension'
	Heyday\MenuManager\MenuSet:
	extensions:
		- 'TractorCow\Fluent\Extension\FluentExtension'
	Heyday\MenuManager\MenuItem:
	extensions:
		- 'TractorCow\Fluent\Extension\FluentExtension'
	SilverStripe\SiteConfig\SiteConfig:
	extensions:
		- 'TractorCow\Fluent\Extension\FluentExtension'
	gorriecoe\Link\Models\Link:
	extensions:
		- 'TractorCow\Fluent\Extension\FluentExtension'

**Comment by tractorcow** [comment]
I suggest not using `alternate_base_url` config in a live site; That config is best used for testing only (override baseurl for testing). It tends to cause unpredictable behaviour in practice, especially on multi-domain sites.

**Comment by tractorcow** [comment]
Ah yes, that would be the cause of the issue :)

Director::host() checks in order:
 - alternate base url
 - request url
 - default base url (from BASE_URL env)

Remove alternate_base_url and it'll work.

**Comment by obj63mc** [comment]
Thanks,  I am trying to work around that.  I currently host on Heroku and with your dynos being hidden by a LB/Proxy, it is always setting the base url to http:// instead of https if I don't set these variables.  Then when forcing SSL you end up in infinite loops and if not set with the base being http:// the CMS admin won't load any of the css assets etc.

Will let you know if I find a workaround for this.

**Comment by obj63mc** [comment]
So finally found a solution - basically due to heroku's environment I know I needed to set the SS_ALLOWED_HOSTS env variable.  The problem is that still wasn't working even though the normal headers are set.  Basically after researching I determined I needed to figure out how to set SS_TRUSTED_PROXY_IPS properly and once set this worked.  

This stack overflow for symfony gave me some direction on what to do.
https://stackoverflow.com/questions/16180486/heroku-trusted-proxies

Basically if you have those two set and remove all the base URL stuff heroku will work as it should and then the domain routing will work as expected.
